### PR TITLE
Multi-architecture dnsmasq container builds + rename container to kub…

### PR DESCRIPTION
**Pull request by girishkalele**
Fixed builds for architecture amd64 - don't have a solution for arm/arm64/ppc yet.

### CODE REVIEW

File: dnsmasq/RELEASES.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Cutting a release
+
+Until we have a proper setup for building this automatically with every binary
+release, here are the steps for making a release. We make releases when they
+are ready, not on every PR.
+
+1. Build the container for testing: 
+
+`cd dnsmasq`
```
Comment (thockin) [inline dnsmasq/RELEASES.md]: you can do multi-line code-blocks as

```
code
goes here
```

File: dnsmasq/RELEASES.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	```
+
+6. Submit a PR for the kubernetes/kubernetes repository to switch to the new TAG version with a do-not-merge label.
+
+7. Manually deploy this to your own cluster by updating the replication
+   controller and deleting the running pod(s).
+
+8. Verify it works.
+
+9. Allow the PR to be merged by removing the do-not-merge label.
```
Comment (thockin) [inline dnsmasq/RELEASES.md]: you don't talk about what to do with changes to *this* repo...

**Comment by girishkalele** [comment]
@luxas 

Tim mentioned that you are familiar with the non-amd64 architecture build processes - if we cannot locate an official alpine-arm base image and dnsmasq-arm binaries, might have to cross-compile dnsmasq from sources.

**Comment by luxas** [comment]
@girishkalele Thanks for caring about multiarch :)
I was just about to fix this, but since you started doing it, I'll review your work and help you get this in.
Will review soon when I have a little more time...

File: dnsmasq/Dockerfile.cross [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM BASEIMAGE
+MAINTAINER Girish Kalele <gkalele@google.com>
+ADD dnsmasq.conf /etc/dnsmasq.conf
+ADD src/dnsmasq /usr/bin/dnsmasq
+ADD DNSMASQVERSION.tar.xz /DNSMASQVERSION.tar.xz
```
Comment (luxas) [inline dnsmasq/Dockerfile.cross]: Why is this included?
It doesn't seem required...

File: dnsmasq/Makefile [hunk 4]
```diff
@@ -20,20 +20,77 @@
...
+CONTAINER_NAME_AND_TAG:=kube-dnsmasq-$(ARCH):$(TAG)
+
+KUBE_CROSS_IMAGE ?= gcr.io/google_containers/kube-cross
+KUBE_CROSS_VERSION ?= v1.6.2-2
+
+DNSMASQ_VERSION ?= dnsmasq-2.76
+
+ifeq ($(ARCH),amd64)
+	BASEIMAGE?=alpine:3.3
+	TRIPLE ?= x86_64-linux-gnu
```
Comment (luxas) [inline dnsmasq/Makefile]: TRIPLE isn't used for amd64

File: dnsmasq/Makefile [hunk 5]
```diff
@@ -20,20 +20,77 @@
...
+endif
 
 all: container
 
-container:
-	# And build the image
-	docker build -t $(REGISTRY)/dnsmasq:$(TAG) .
+sources:
+	@echo Downloading dnsmasq sources for cross compiled architectures
+	wget http://www.thekelleys.org.uk/dnsmasq/$(DNSMASQ_VERSION).tar.xz && tar xf $(DNSMASQ_VERSION).tar.xz
```
Comment (luxas) [inline dnsmasq/Makefile]: `curl -sSL http://www.thekelleys.org.uk/dnsmasq/$(DNSMASQ_VERSION).tar.xz | tar -xz` is cleaner, and doesn't leave `$(DNSMASQ_VERSION).tar.xz` hanging around

File: dnsmasq/Makefile [hunk 6]
```diff
@@ -20,20 +20,77 @@
...
 
-push: container
+container.arm: sources container.cross
```
Comment (luxas) [inline dnsmasq/Makefile]: I don't think sources is necessary here, as container.cross depends on it

File: dnsmasq/Makefile [hunk 7]
```diff
@@ -20,20 +20,77 @@
...
+
+container.arm64: sources container.cross
+
+container.ppc64le: sources container.cross
+
+container.cross: sources
+	# Build the dnsmasq binary using the cross-compilation docker image
+	docker run --detach=false --rm -v `pwd`/$(DNSMASQ_VERSION):/build $(KUBE_CROSS_IMAGE):$(KUBE_CROSS_VERSION) bash -c "cd /build; make clean; CC=$(TRIPLE)-gcc make"
```
Comment (luxas) [inline dnsmasq/Makefile]: `--detach=false` => `-it`
consider merging `$(KUBE_CROSS_IMAGE):$(KUBE_CROSS_VERSION)` to only one variable, to make it less verbose

File: dnsmasq/Makefile [hunk 8]
```diff
@@ -20,20 +20,77 @@
...
+	cp Dockerfile.cross $(DNSMASQ_VERSION)/Dockerfile
+	sed -i s:BASEIMAGE:$(BASEIMAGE):g $(DNSMASQ_VERSION)/Dockerfile
```
Comment (luxas) [inline dnsmasq/Makefile]: please `sed -i "s|BASEIMAGE|$(BASEIMAGE)|g"`, to make it more in line with the other scripts

File: dnsmasq/Makefile (hunk 5 above)
Comment (luxas) [inline dnsmasq/Makefile]: Also, why `thekelleys.org.uk`?

Comment (girishkalele) [inline dnsmasq/Makefile]: Looks like that is where the sources are.

http://www.thekelleys.org.uk/dnsmasq/doc.html

Simon Kelley is the author and maintainer.

File: dnsmasq/Dockerfile.cross (hunk 3 above)
Comment (girishkalele) [inline dnsmasq/Dockerfile.cross]: I wanted to include the sources used to build it, don't know the normal practice for building containers - I will remove them.

**Comment by luxas** [comment]
My main question is, does it work?
Unfortunately, I think it won't. You may easily test that by "cross"-compiling for `amd64` too, just change the `container.amd64` rule to `container.cross`

**Comment by luxas** [comment]
@girishkalele I think I have a solution to this problem, experimented a bit with it.
I'll upload a related PR to this one that builds another image when I have time

**Comment by girishkalele** [comment]
@luxas 

I see you meant whether the sources are going to work - I didn't try dnsmasq from these sources - the amd64 version we will still get from Alpine Linux apk repo.
Is your related PR is adding on to my changes ?

**Comment by luxas** [comment]
See: #1146

**Comment by luxas** [comment]
I would try to keep the Makefile more simple, like this:
```
TAG?=1.2
REGISTRY?=gcr.io/google_containers
ARCH?=amd64
KUBE_CROSS_IMAGE:=$(REGISTRY)/kube-cross:v1.6.2-2
DNSMASQ_VERSION?=dnsmasq-2.76
TEMP_DIR:=$(shell mktemp -d)

ifeq ($(ARCH),amd64)
	BASEIMAGE?=alpine:3.3
else
	BASEIMAGE?=$(REGISTRY)/tiny-glibc-$(ARCH):1.0
endif

ifeq ($(ARCH),arm)
	TRIPLE?=arm-linux-gnueabi
endif
ifeq ($(ARCH),arm64)
	TRIPLE?=aarch64-linux-gnu
endif
ifeq ($(ARCH),ppc64le)
	TRIPLE?=powerpc64le-linux-gnu
endif

all: build

build:
ifeq ($(ARCH),amd64)
	cp Dockerfile $(TEMP_DIR)
else
	cp Dockerfile.cross $(TEMP_DIR)/Dockerfile
	sed -i "s|BASEIMAGE|$(BASEIMAGE)|g" $(TEMP_DIR)/Dockerfile

	# Build the dnsmasq binary using the cross-compilation docker image
	docker run -it --rm -v $(TEMP_DIR):/build $(KUBE_CROSS_IMAGE) /bin/bash -c \
		"curl -sSL  http://www.thekelleys.org.uk/dnsmasq/$(DNSMASQ_VERSION).tar.xz | tar -xJ \
		&& cd $(DNSMASQ_VERSION) \
		&& CC=$(TRIPLE)-gcc make \
		&& cp src/dnsmasq /build"
endif
	
	cp dnsmasq.conf $(TEMP_DIR)
	docker build -t $(REGISTRY)/dnsmasq-$(ARCH):$(TAG) $(TEMP_DIR)


push: build
ifeq ($(REGISTRY),gcr.io/google_containers)
	gcloud docker push $(REGISTRY)/dnsmasq-$(ARCH):$(TAG)
else
	docker push $(REGISTRY)/dnsmasq-$(ARCH):$(TAG)
endif
```
and
```
FROM BASEIMAGE
MAINTAINER Girish Kalele <gkalele@google.com>
COPY dnsmasq.conf /etc/dnsmasq.conf
COPY dnsmasq /usr/bin/dnsmasq
ENTRYPOINT ["/usr/bin/dnsmasq", "--keep-in-foreground"]
```

Feel free to copy the code if you like, or I can send a new PR

**Comment by girishkalele** [comment]
@luxas 

Integrated your changes, testing now - I need to pull your request #1146 and build those glibc containers locally.

File: dnsmasq/Dockerfile.cross [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM BASEIMAGE
+MAINTAINER Girish Kalele <gkalele@google.com>
+COPY dnsmasq.conf /etc/dnsmasq.conf
+COPY dnsmasq /usr/bin/dnsmasq
```
Comment (luxas) [inline dnsmasq/Dockerfile.cross]: Saw this now, maybe it's best to align with amd64, i.e. `/usr/sbin/dnsmasq`...

File: dnsmasq/Dockerfile [hunk 10]
```diff
@@ -16,4 +16,4 @@ FROM alpine:3.3
 MAINTAINER Abhishek Shah <abshah@google.com>
 RUN apk --no-cache add dnsmasq
 ADD dnsmasq.conf /etc/dnsmasq.conf
```
Comment (luxas) [inline dnsmasq/Dockerfile]: nit: I think `COPY` is preferred over `ADD` in this case

File: dnsmasq/Changelog [hunk 11]
```diff
@@ -0,0 +1,4 @@
+## Version 1.1 (Fri May 27 2016 Abhi S. <abshah@google.com>
```
Comment (luxas) [inline dnsmasq/Changelog]: Update to 1.2

File: dnsmasq/Makefile [hunk 12]
```diff
@@ -20,20 +20,54 @@
...
+TAG?=1.2
 REGISTRY?=gcr.io/google_containers
+ARCH?=amd64
+KUBE_CROSS_IMAGE:=$(REGISTRY)/kube-cross:v1.6.2-2
+DNSMASQ_VERSION?=dnsmasq-2.76
+TEMP_DIR:=$(shell mktemp -d)
 
-all: container
+ifeq ($(ARCH),amd64)
+    BASEIMAGE?=alpine:3.3
```
Comment (luxas) [inline dnsmasq/Makefile]: I think 3.4 is out

File: dnsmasq/RELEASES.md (hunk 1 above)
Comment (luxas) [inline dnsmasq/RELEASES.md]: \``` console

$ cd dnsmasq
$ make build REGISTRY={your-name} TAG=rc ARCH={your-architecture}

\```

like @thockin said

File: dnsmasq/RELEASES.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+4. Update the TAG version in `Makefile` and update the `Changelog`.
+
+5. Build and push the container for real for all architectures:
+
+	```console
+	# Build for linux/amd64 (default)
+	$ make push ARCH=amd64
+	# ---> gcr.io/google_containers/kube-dnsmasq-amd64:TAG
+	# ---> gcr.io/google_containers/kube-dnsmasq:TAG (image with backwards-compatible naming)
```
Comment (luxas) [inline dnsmasq/RELEASES.md]: We don't do this with this image and shouldn't, so remove this line

**Comment by luxas** [comment]
Some small things, then LGTM
